TE-10560 Documentation page for Spryker Elastic Computing (#934)
spryker/spryker-docs: docs/scos/dev/guidelines/performance-guidelines/spryker-elastic-computing/spryker-elastic-computing.md

---
title: Spryker Elastic Computing
description: This guideline explains how to use Elastic Computing functionality.
last_updated: May 16, 2022
template: concept-topic-template
---

Spryker provides several elastic computing features OOTB that allows to utilise computational resources in a more efficient way. 

# Monitoring
For enabling New Relic monitoring for P&S and queue workers infrastructure, please follow the next steps:

1. Update `spryker/monitoring` module at least to v2.5.0.
2. Add `queue:task:start` command to the argument grouped transactions in `Pyz\Zed\Monitoring\MonitoringConfig` configuration class
```php
namespace Pyz\Zed\Monitoring;

use Spryker\Zed\Monitoring\MonitoringConfig as BaseMonitoringConfig;

class MonitoringConfig extends BaseMonitoringConfig
{
    /**
     * @api
     *
     * @return array<string>
     */
    public function getGroupedByArgumentTransactions(): array
    {
        return ['queue:task:start'];
    }
}
```

> 3. Add to the `Pyz\Zed\Monitoring\Business\MonitoringBusinessFactory` following lines of code:

Reviewer: Explain what this does

Reviewer 2: Enabling `FirstArgumentMonitoringConsoleTransactionNamingStrategy` to be used for transaction naming.

```php
public function getMonitoringTransactionNamingStrategies(): array
    {
        return [
            return new FirstArgumentMonitoringConsoleTransactionNamingStrategy($this->getConfig()->getGroupedByArgumentTransactions());
        ];
    }
```

> This will aggregates all different queue worker transactions under corresponding `queue:task:start [queue_name]` transaction name in New Relic.

Reviewer: ```suggestion
Now New Relic aggregates all queue worker transactions under the name specified in `queue:task:start {queue_name}`
```

As a result, the transactions will appear in New Relic dashboard as on the following illustration:

# RAM-aware batch processing integration
Batch processing allows significantly speed up processing operations. Moving more data into RAM at once allows decrease the number of I/O thus winning some operations wall time. To win even more time one can calculate batch size dynamically, based on available RAM. 

> This can be achieved with elastic batch concept introduced in `spryker/data-import` module. The following instruction will explain the integration details of elastic batch using the glossary data import as an example.

Reviewer: As I understand, the module is responsible for data import in general. Did you just add batch processing to the `spryker/data-import`?

---
title: Spryker Elastic Computing
description: This guideline explains how to use Elastic Computing functionality.
last_updated: May 16, 2022
template: concept-topic-template
---

Spryker is shipped with several elastic computing features that let you to utilize computational resources more efficiently. 

## Integrate New Relic monitoring

> To integrate New Relic monitoring for publish and sync (P&S) and the infastructure of queue workers, follow the steps:

Reviewer: `infastructure of queue workers` is part of `publish and sync (P&S), isn't it?

last_updated: May 16, 2022
template: concept-topic-template
---

Spryker is shipped with several elastic computing features that let you to utilize computational resources more efficiently. 

## Monitoring

To enable New Relic monitoring for P&S and the infastructure of queue workers, follow the steps:

1. Update the `spryker/monitoring` module to version 2.5.0 or higher.
2. In `Pyz\Zed\Monitoring\MonitoringConfig`, add `queue:task:start` command to the argument grouped transactions:
```php
namespace Pyz\Zed\Monitoring;

use Spryker\Zed\Monitoring\MonitoringConfig as BaseMonitoringConfig;

class MonitoringConfig extends BaseMonitoringConfig
{
    /**
     * @api
     *
     * @return array<string>
     */
    public function getGroupedByArgumentTransactions(): array
    {
        return ['queue:task:start'];
    }
}
```

3. Add to the `Pyz\Zed\Monitoring\Business\MonitoringBusinessFactory` following lines of code:
```php
public function getMonitoringTransactionNamingStrategies(): array
    {
        return [
            return new FirstArgumentMonitoringConsoleTransactionNamingStrategy($this->getConfig()->getGroupedByArgumentTransactions());
        ];
    }
```

> This will aggregate all different queue worker transactions under corresponding `queue:task:start {queue_name}` transaction name in New Relic.

Reviewer: ```suggestion
Now New Relic aggregates all queue worker transactions under the name specified in `queue:task:start {queue_name}`.
```

Reviewer 2: Now New Relic transactions of `queue:task:start` are aggregated under separate names for each queue name.

> :info: Similar integration must be applied for all project data importers.

1. In `Pyz\Zed\DataImport\DataImportConfig`, add the following configuration:

* `BulkWriteGradualityFactor`: estimates an upper limit of memory that can be safely utilized by operations. A bigger value enables a more precise approximation but requires more iterations. A smaller value gives a less precise approximation but requires less iterations.

* `BulkWriteMemoryThresholdPercent`: defines a margin of PHP memory limit configured in PHP.

```php
/**
 * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
 * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
 * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
 */
class DataImportConfig extends SprykerDataImportConfig
{
    ...
        
    /**
     * @return int
     */
    public function getBulkWriteGradualityFactor(): int
    {
        return 5;
    }

    /**
     * @return int
     */
    public function getBulkWriteMemoryThesoldPercent(): int
    {
        return 95;
    }
}
```

2. In `src/Pyz/Zed/DataImport/Business/DataImportBusinessFactory.php`, adjust the creation of data importer so it’s new version will

> - use `DataSetStepBrokerElasticBatchTransactionAware` data set step broker and

Reviewer: What is a `data set step broker`?

Author: This is a class (in our case) which is responsible for wrapping out a logic related to the processing of one data set (one row from the file with data).

- the writer steps receive the `MemoryAllocatedElasticBatch`
```php
/**
 * @method \Pyz\Zed\DataImport\DataImportConfig getConfig()
 * @SuppressWarnings(PHPMD.CyclomaticComplexity)
 * @SuppressWarnings(PHPMD.ExcessiveClassComplexity)
 * @SuppressWarnings(PHPMD.ExcessiveClassLength)
 */
class DataImportBusinessFactory extends SprykerDataImportBusinessFactory
{
  ...
  protected function createGlossaryImporter(
        DataImportConfigurationActionTransfer $dataImportConfigurationActionTransfer 
        ): DataImporterInterface
    {
      $dataImporter = $this->getCsvDataImporterFromConfig(
            $this->getConfig()->buildImporterConfigurationByDataImportConfigAction($dataImportConfigurationActionTransfer),
        );
      $dataSetStepBroker = $this->createElasticBatchTransactionAwareDataSetStepBroker(GlossaryWriterStep::BULK_SIZE);
      $dataSetStepBroker
        ->addStep($this->createLocaleNameToIdStep(GlossaryWriterStep::KEY_LOCALE))
        ->addStep(new GlossaryWriterStep($this->createMemoryAllocatedElasticBatch()));

        $dataImporter->addDataSetStepBroker($dataSetStepBroker);

        return $dataImporter;
    }
  ...
}
```

> 3. Update the writer steps (`Pyz\Zed\DataImport\Business\Model\Glossary\GlossaryWriterStep`) as it is show in the code below by adjusting the execute method to flush `MemoryAllocatedElasticBatch` when it is full.

Reviewer: Why is `Pyz\Zed\DataImport\Business\Model\Glossary\GlossaryWriterStep` in paratheses? Is it an example?

Reviewer 2: Yes, this is an example. We are specifying the actual file to be changed with the following block of code.

            ->filterByFkLocale($dataSet[static::KEY_ID_LOCALE])
            ->findOneOrCreate();
        $glossaryTranslationEntity
            ->setValue($dataSet[static::KEY_TRANSLATION])
            ->save();
        $this->addPublishEvents(GlossaryStorageConfig::GLOSSARY_KEY_PUBLISH_WRITE, $glossaryTranslationEntity->getFkGlossaryKey());
    }
}
```

# Scalable application infrastructure for P&S workers
1. Update `spryker/queue` module at least to v1.10.0.
2. Configure `THREAD_POOL_SIZE` parameter equals to the number of your CPU cores plus one.
```php
  <?php

namespace Pyz\Zed\Queue;

use Generated\Shared\Transfer\RabbitMqConsumerOptionTransfer;
use Spryker\Shared\Config\Config;
use Spryker\Shared\Event\EventConstants;
use Spryker\Shared\Log\LogConstants;
use Spryker\Shared\Queue\QueueConstants;
use Spryker\Zed\Queue\QueueConfig as SprykerQueueConfig;

class QueueConfig extends SprykerQueueConfig
{
  public const CPU_CORE_NUMBER = 32; // your system's CPU core count
 
    /**
     * @return array<string, mixed>
     */
    protected function getThreadPoolSize(): int
    {
        return static::CPU_CORE_NUMBER + 1;
    }
 
}
```

> As the result, worker will spawn a group of processes per each non empty queue with consideration of number of messages and available RAM. 

Reviewer: Is it only one worker or there are multiple workers that spawn processes?

            ->filterByFkLocale($dataSet[static::KEY_ID_LOCALE])
            ->findOneOrCreate();
        $glossaryTranslationEntity
            ->setValue($dataSet[static::KEY_TRANSLATION])
            ->save();
        $this->addPublishEvents(GlossaryStorageConfig::GLOSSARY_KEY_PUBLISH_WRITE, $glossaryTranslationEntity->getFkGlossaryKey());
    }
}
```

## Integrate scalable application infrastructure for P&S workers

1. Update module `spryker/queue` to version 1.10.0 or higher. 
2. Set `THREAD_POOL_SIZE` to the number of your CPU cores plus one.
```php
  <?php

namespace Pyz\Zed\Queue;

use Generated\Shared\Transfer\RabbitMqConsumerOptionTransfer;
use Spryker\Shared\Config\Config;
use Spryker\Shared\Event\EventConstants;
use Spryker\Shared\Log\LogConstants;
use Spryker\Shared\Queue\QueueConstants;
use Spryker\Zed\Queue\QueueConfig as SprykerQueueConfig;

class QueueConfig extends SprykerQueueConfig
{
  public const CPU_CORE_NUMBER = 32; // your system's CPU core count
 
    /**
     * @return array<string, mixed>
     */
    protected function getThreadPoolSize(): int
    {
        return static::CPU_CORE_NUMBER + 1;
    }
 
}
```

> As a result, worker spawn a group of processes per each non-empty queue based on number of messages and available RAM. 

Reviewer: Is it only one or multiple workers spawning processes?
